components/org.wso2.carbon.identity.organization.management.organization.user.sharing/src/main/java/org/wso2/carbon/identity/organization/management/organization/user/sharing/UserSharingPolicyHandlerServiceImpl.java

```
@@ -294,39 +278,175 @@ public ResponseSharedRolesDO getRolesSharedWithUserInOrganization(String associa
+        try {
+            startTenantFlowFromOrganization(sharingInitiatedOrgId);
+            for (SelectiveUserShareOrgDetailsDO organization : validOrganizations) {
+                populateSelectiveUserShareByCriteria(organization, userCriteria, sharingInitiatedOrgId);
+            }
+            LOG.info("Completed user selective share initiated from " + getOrganizationManager() + ".");
+        } finally {
+            PrivilegedCarbonContext.endTenantFlow();
+        }
+    }
+
+    private void processGeneralUserShare(Map<String, UserCriteriaType> userCriteria, PolicyEnum policy,
+                                         List<String> roleIds, String sharingInitiatedOrgId) {
+
+        try {
+            startTenantFlowFromOrganization(sharingInitiatedOrgId);
+            for (Map.Entry<String, UserCriteriaType> criterion : userCriteria.entrySet()) {
+                String criterionKey = criterion.getKey();
+                UserCriteriaType criterionValues = criterion.getValue();
+
+                try {
+                    switch (criterionKey) {
+                        case USER_IDS:
+                            if (criterionValues instanceof UserIdList) {
+                                generalUserShareByUserIds((UserIdList) criterionValues, policy, roleIds,
+                                        sharingInitiatedOrgId);
+                            } else {
+                                LOG.error("Invalid user criteria provided for general user share: " + criterionKey);
+                            }
+                            break;
+                        default:
+                            LOG.error("Invalid user criteria provided for general user share: " + criterionKey);
+                    }
+                } catch (UserSharingMgtException e) {
+                    LOG.debug("Error occurred while sharing user from user criteria: " + USER_IDS, e);
```

Need to throw an exception

---

Since this is an asynchronous operation, directly throwing an exception could cause issues. For instance, `CompletableFuture` does not implicitly allow handled exceptions to be thrown. Additionally, since the API response has already been delivered at this point, there is no direct way to propagate the error back to the user. Therefore, logging the error would be more appropriate in this case. Moreover, we can later trigger event handlers to indicate that these operations have failed in a structured manner rather than relying solely on logging.